aws/rust-runtime/aws-inlineable/src/http_body_checksum.rs

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Functions for modifying requests and responses for the purposes of checksum validation
+
+/// Errors related to constructing checksum-validated HTTP requests
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) enum Error {
+    /// Only request bodies with a known size can be checksum validated
+    UnsizedRequestBody,
+    ChecksumHeadersAreUnsupportedForStreamingBody,
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::UnsizedRequestBody => write!(
+                f,
+                "Only request bodies with a known size can be checksum validated."
+            ),
+            Self::ChecksumHeadersAreUnsupportedForStreamingBody => write!(
+                f,
+                "Checksum header insertion is only supported for non-streaming HTTP bodies. \
+                 To checksum validate a streaming body, the checksums must be sent as trailers."
+            ),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+/// Given a `&mut http::request::Request` and a `aws_smithy_checksums::ChecksumAlgorithm`,
+/// calculate a checksum and modify the request to include the checksum as a header.
+///
+/// **NOTE: This function only supports non-streaming request bodies and will return an error if a
+/// streaming request body is passed.**
+#[allow(dead_code)]
+pub(crate) fn calculate_body_checksum_and_insert_as_header(
+    request: &mut http::request::Request<aws_smithy_http::body::SdkBody>,
+    checksum_algorithm: aws_smithy_checksums::ChecksumAlgorithm,
+) -> Result<(), aws_smithy_http::operation::BuildError> {
+    let mut checksum = checksum_algorithm.into_impl();
+    match request.body().bytes() {
+        Some(data) => {
+            checksum.update(data);
+
+            request
+                .headers_mut()
+                .insert(checksum.header_name(), checksum.header_value());
+
+            Ok(())
+        }
+        None => Err(aws_smithy_http::operation::BuildError::Other(Box::new(
+            Error::ChecksumHeadersAreUnsupportedForStreamingBody,
+        ))),
+    }
+}
+
+/// Given a `&mut http::request::Request` and a `aws_smithy_checksums::ChecksumAlgorithm`, wrap the
+/// request's inner body with a body that will calculate a checksum during streaming and emit the
+/// checksum as a trailer. This only supports streaming bodies with a known size and will otherwise
+/// return an error.
+#[allow(dead_code)]
+pub(crate) fn wrap_streaming_request_body_in_checksum_calculating_body(
```

Comment:
should we make these private and make a root function that always just does the right thing based on the output of request.bod().bytes()?